osu.Game.Tests/Online/TestSceneMultiplayerBeatmapTracker.cs

```diff
@@ -0,0 +1,181 @@
+
+            AddStep("set progress 40%", () => ((TestDownloadRequest)beatmaps.GetExistingDownload(testBeatmapSet)).SetProgress(0.4f));
+            addAvailabilityCheckStep("state downloading 40%", () => BeatmapAvailability.Downloading(0.4f));
+
+            AddStep("finish download", () => ((TestDownloadRequest)beatmaps.GetExistingDownload(testBeatmapSet)).TriggerSuccess(testBeatmapFile));
+            addAvailabilityCheckStep("state importing", BeatmapAvailability.Importing);
+
+            AddStep("allow importing", () => beatmaps.AllowImport.SetResult(true));
+            AddUntilStep("wait for import", () => beatmaps.IsAvailableLocally(testBeatmapSet));
+            addAvailabilityCheckStep("state locally available", BeatmapAvailability.LocallyAvailable);
+        }
+
+        [Test]
+        public void TestTrackerRespectsSoftDeleting()
+        {
+            AddStep("allow importing", () => beatmaps.AllowImport.SetResult(true));
+            AddStep("import beatmap", () => beatmaps.Import(testBeatmapSet).Wait());
+            addAvailabilityCheckStep("state locally available", BeatmapAvailability.LocallyAvailable);
+
+            AddStep("delete beatmap", () => beatmaps.Delete(testBeatmapSet));
+            addAvailabilityCheckStep("state not downloaded", BeatmapAvailability.NotDownloaded);
+
+            AddStep("undelete beatmap", () => beatmaps.Undelete(testBeatmapSet));
+            addAvailabilityCheckStep("state locally available", BeatmapAvailability.LocallyAvailable);
+        }
+
+        [Test]
+        public void TestTrackerRespectsChecksum()
+        {
+            AddStep("allow importing", () => beatmaps.AllowImport.SetResult(true));
+
+            BeatmapInfo wrongBeatmap = null;
+
+            AddStep("import wrong checksum beatmap", () =>
+            {
+                wrongBeatmap = new TestBeatmap(Ruleset.Value).BeatmapInfo;
+                wrongBeatmap.MD5Hash = "1337";
+
+                beatmaps.Import(wrongBeatmap.BeatmapSet).Wait();
+            });
+            AddAssert("wrong beatmap available", () => beatmaps.QueryBeatmap(b => b.OnlineBeatmapID == wrongBeatmap.OnlineBeatmapID) != null);
+            addAvailabilityCheckStep("state still not downloaded", BeatmapAvailability.NotDownloaded);
+
+            AddStep("recreate tracker", () => Child = tracker = new MultiplayerBeatmapTracker
+            {
+                SelectedItem = { BindTarget = selectedItem }
+            });
+            addAvailabilityCheckStep("state not downloaded as well", BeatmapAvailability.NotDownloaded);
+        }
+
+        private void addAvailabilityCheckStep(string description, Func<BeatmapAvailability> expected)
+        {
+            // In DownloadTrackingComposite, state changes are scheduled one frame later, wait one step.
+            AddWaitStep("wait for potential change", 1);
+            AddAssert(description, () => tracker.Availability.Value.Equals(expected.Invoke()));
```

Comment:
I intentionally didn't because really all that matters is waiting for one frame, which will always be achieved with adding a waiting step, as in the above comment.

Reply:
...what? you do realise a wait step is not one frame? surely an until step is more reliable?

Reply:
perhaps I haven't explained properly, adding a wait step will delays the assert from being executed, to one frame next at most.

an until step will keep waiting for several frames if the tracker for some reason did not change availability in time and pass the test, even though that's not correct, so this way will be more accurate in terms of testing.

Reply:
This fails randomly still.

```
osu.Game.Tests.Online.TestSceneMultiplayerBeatmapAvailabilityTracker.TestBeatmapDownloadingFlow

TearDown : osu.Framework.Testing.Drawables.Steps.AssertButton+TracedException : state locally available 
--TearDown
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Platform.HeadlessGameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.ProcessFrame()
   at osu.Framework.Threading.GameThread.runWork()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()




-----

One or more child tests had errors
  Exception doesn't have a stacktrace



34414733: Running MultiplayerBeatmapAvailabilityTracker (TestSceneMultiplayerBeatmapAvailabilityTracker) visual test cases...
34414733: TestSceneMultiplayerBeatmapAvailabilityTracker
34414933: step 1 Assert: ensure beatmap unavailable
34415133: step 2 Repeat: wait for potential change 1/1
34415333: step 3 Assert: state not downloaded
34415533: step 4 start downloading
34415733: step 5 Repeat: wait for potential change 1/1
34415933: step 6 Assert: state downloading 0%
34416133: step 7 set progress 40%
34416333: step 8 Repeat: wait for potential change 1/1
34416533: step 9 Assert: state downloading 40%
34416733: step 10 finish download
34416933: step 11 Repeat: wait for potential change 1/1
34417133: step 12 Assert: state importing
34417333: step 13 allow importing
34417533: step 14 Repeat: wait for import (1 tries)..............................................[database] 2021-02-01 08:39:47 [verbose]: [?????] Beginning import...
..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................[database] 2021-02-01 08:39:48 [verbose]: [672a6] Validating online IDs for 12 beatmaps...
......................................................................................................................................................................................................................................................................................
[database] 2021-02-01 08:39:48 [verbose]: [672a6] Import successfully completed!
34616533: step 15 Repeat: wait for potential change 1/1
```

Reply:
Right, I got overthrown by this near-impossible possibility that the `State` could be changed after the assert step is called because parent's scheduler runs before children. 

Turns out what the issue actually was is that the "wait for import" until step was relying on `IsAvailableLocally()`, so given that the import is performed in another thread, this method may return `true` early before the importing has finished and `ItemUpdated` is invoked. 

So I fixed it by retrieving the current import task and waiting for it until it completely finishes, and have made the test go through this "run selected tests until failure" feature for around an hour, with no sign of failure!